packages/repco-frontend/app/components/player/Player.tsx

```diff
@@ -0,0 +1,422 @@
+import type { Track } from './usePlaylists'
+import { Button } from '../ui/primitives/Button'
+
+type PlayerContext = {
+  track: Track | null
+  setTrack: React.Dispatch<React.SetStateAction<Track | null>>
+  trackIndex: number
+  setTrackIndex: React.Dispatch<React.SetStateAction<number>>
+  nextTrack: () => void
+  previousTrack: () => void
+  tracks: Track[]
+  queueVisibility: boolean
+  setQueueVisibility: React.Dispatch<React.SetStateAction<boolean>>
+}
+
+type PlaystateContext = {
+  audio: HTMLAudioElement | null
+  isPlaying: boolean
+  isReady: boolean
+  currentTime: number
+  duration: number
+}
+
+/**
+ * The player context holds setters and values for the
+ * currently playing media track.
+ */
+export const PlayerContext = React.createContext<PlayerContext | undefined>(
+  undefined,
+)
+
+/**
+ * The playstate context holds the audio element and play state (time, play/pause, ...)
+ */
+const PlaystateContext = React.createContext<PlaystateContext | undefined>(
+  undefined,
+)
+
+export function PlayTrackButton({ track }: { track: Track }) {
+  const { tracks, replaceCurrentQueue } = useQueue()
+  const player = usePlayer()
+
+  const clickHandler = () => {
+    const currentIndex = player?.trackIndex || 0
+    if (tracks.find((item) => track.uid === item.uid)) {
+      const index = tracks.findIndex((item) => track.uid === item.uid)
+      player?.setTrackIndex(index)
+    } else {
+      if (tracks.length === 0) {
+        replaceCurrentQueue([track])
+        player?.setTrackIndex(0)
+      } else {
+        tracks.splice(currentIndex + 1, 0, track)
+        replaceCurrentQueue(tracks)
+        player?.setTrackIndex(currentIndex + 1)
+      }
+    }
+  }
+
+  return (
+    <Button onClick={clickHandler}>
+      <PlayIcon />
+    </Button>
+  )
+}
+
+/**
+ * The player provider provides the player and playstate contexts and renders an (invisible) audio element.
+ * It also implements basic logic: Set audio element src on track change, change position on mark change, etc.
+ */
+export function PlayerProvider({ children }: PropsWithChildren) {
+  const [didMount, setDidMount] = useState(false)
+  const { tracks } = useQueue()
+  const [track, setTrack] = useState<Track | null>(tracks.at(0) || null)
+  const [trackIndex, setTrackIndex] = useState(0)
+  const [queueVisibility, setQueueVisibility] = useState(false)
+
+  const src = track ? track.src : null
+  const { audio, element, ...state } = useAudioElement({ src })
+
+  useEffect(() => {
+    if (tracks.at(trackIndex)) {
+      setTrack(tracks.at(trackIndex) || null)
+    } else if (tracks.length === 0 && audio) {
+      setTrack(null)
+      audio.src = ''
+    }
+  }, [tracks, trackIndex])
+
+  useEffect(() => {
+    if (!audio) return
+    if (!track) {
+      audio.src = ''
+      return
+    }
+    let pos = 0
+    audio.currentTime = pos
+    if (didMount) {
+      audio.play()
+    }
+    setDidMount(true)
+  }, [audio, track])
```

Comment:
I think the dependencies should include `didMount`?

Reply:
I don't understand exactly what you mean? didMount only helps here that the player is not rendered during initial rendering.